include/deal.II/fe/fe_poly.h

```diff
@@ -420,29 +429,24 @@ class FE_Poly : public FiniteElement<dim,spacedim>
     std::vector<std::vector<Tensor<2,dim> > > shape_hessians;
 
     /**
-     * Scratch array to store temporary values during hessian calculations in
-     * actual cells.
+     * Array with shape function third derivatives in quadrature points. There
+     * is one row for each shape function, containing values for each
+     * quadrature point.
+     *
+     * We store the third derivatives in the quadrature points on the unit
+     * cell. We then only have to apply the transformation when visiting an
+     * actual cell.
      */
-    mutable std::vector<Tensor<2,dim> > untransformed_shape_hessians;
+    std::vector<std::vector<Tensor<3,dim> > > shape_3rd_derivatives;
   };
 
-  /**
-   * Correct the hessian in the reference cell by subtracting the term corresponding
-   * to the Jacobian gradient for one degree of freedom. The result being given by:
-   * @f[
-   * \frac{\partial^2 \phi_i}{\partial\hat{x}_J\partial\hat{x}_K}
-   * - \frac{\partial \phi_i}{\partial {x}_l}
-   * \left( \frac{\partial^2{x}_l}{\partial\hat{x}_J\partial\hat{x}_K} \right).
-   * @f]
-   * After this correction, the shape hessians are simply a mapping_covariant_gradient
-   * transformation.
-   */
+// Correct the shape third derivatives by subtracting the terms corresponding
+// to the Jacobian pushed forward gradient and second derivative.
```

Comment:
I think it would be good if you left a doxygen-style comment here that spells out the formula for what this function does.